File: agent/lib/conventions.ts
/**
 * Dynamic Research conventions — TypeScript wrapper.
 *
 * Canonical data lives in agent/lib/conventions.json. This module loads
 * it at runtime, narrows types, and exposes idiomatic helpers. Don't
 * duplicate values here — change conventions.json, the wrapper picks up.
 *
 * Usage:
 *   import { slugify, studioFilename, isSkipFile, RESEARCH_ROLES } from "../lib/conventions.js";
 */

import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";

// ── Load canonical data ─────────────────────────────────────────────

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const CONVENTIONS_PATH = path.join(__dirname, "conventions.json");

interface SlugifyConfig {
  max_length: number;
  strip_pattern_js: string;
  fallback: string;
}

interface ProductDef {
  ext: string;
  list_method: string;
  download_method: string;
}

interface FilenamePatterns {
  studio: { pattern: string; regex: string; products: Record<string, ProductDef> };
  research: { pattern: string; regex: string; roles: string[] };
  research_docx_companion: { pattern: string; regex: string; roles: string[] };
}

interface SkipFiles {
  exact: string[];
  prefixes: string[];
  extensions?: string[];
}

/**
 * v3 (S102) user file-upload feature. Caps + path-segment names for
 * user-attached source files. Pair-edited frontend mirror:
 * frontend/lib/attachments-constants.ts.
 */
interface AttachmentsConfig {
  allowed_extensions: string[];
  allowed_mime_types: string[];
  /** Pattern every storedName must match; compiled by storage-paths.ts. */
  stored_name_regex: string;
  max_file_bytes: number;
  max_total_bytes: number;
  max_files: number;
  staging_prefix: string;
  sources_subdir: string;
  staging_ttl_hours: number;
  max_pages_read_per_pdf: number;
  max_digest_words_per_file: number;
}

interface Conventions {
  _version: number;
  _last_updated: string;
  slugify: SlugifyConfig;
  filename_patterns: FilenamePatterns;
  skip_files: SkipFiles;
  attachments: AttachmentsConfig;
  supabase_storage: { bucket: string };
}

const raw = fs.readFileSync(CONVENTIONS_PATH, "utf-8");
const data = JSON.parse(raw) as Conventions;

// ── Public API ──────────────────────────────────────────────────────

export const VERSION = data._version;
export const LAST_UPDATED = data._last_updated;
export const BUCKET = data.supabase_storage.bucket;

export const SKIP_FILES = new Set(data.skip_files.exact);
export const SKIP_PREFIXES = data.skip_files.prefixes;
export const SKIP_EXTENSIONS = data.skip_files.extensions ?? [];

/**
 * S102 file-upload caps + path segments. Consumed by agent/lib/storage-paths.ts
 * (staging/sources path helpers), the executor's attachment download, and the
 * staging TTL sweep. conventions.json is the canonical source; the frontend
 * duplicates these values in frontend/lib/attachments-constants.ts (pair-edit).
 */
export const ATTACHMENTS = data.attachments;

export const RESEARCH_ROLES = new Set(data.filename_patterns.research.roles);
export const RESEARCH_DOCX_ROLES = new Set(data.filename_patterns.research_docx_companion.roles);
export const STUDIO_PRODUCTS = data.filename_patterns.studio.products;

export const STUDIO_FILENAME_REGEX = new RegExp(data.filename_patterns.studio.regex);
export const RESEARCH_FILENAME_REGEX = new RegExp(data.filename_patterns.research.regex);
export const RESEARCH_DOCX_FILENAME_REGEX = new RegExp(data.filename_patterns.research_docx_companion.regex);

/** Canyon Lake S12 slugify: strip special, spaces->hyphens, lowercase, max length. */
export function slugify(title: string | null | undefined): string {
  if (!title) return data.slugify.fallback;
  const stripPattern = new RegExp(data.slugify.strip_pattern_js, "g");
  let s = title.replace(stripPattern, "");
  s = s.trim().replace(/\s+/g, "-");
  s = s.replace(/-+/g, "-").toLowerCase();
  s = s.slice(0, data.slugify.max_length).replace(/-+$/, "");
  return s || data.slugify.fallback;
}

/** Build a Studio product filename: {title-slug}-{TIMESTAMP}-{product}.{ext} */
export function studioFilename(title: string, timestamp: string, product: string): string {
  const def = STUDIO_PRODUCTS[product];
  if (!def) throw new Error(`unknown studio product: ${product}`);
  return `${slugify(title)}-${timestamp}-${product}.${def.ext}`;
}

/** Build a research file filename: {topic-prefix}-{role}.{ext} */
export function researchFilename(topicPrefix: string, role: string, ext: string): string {
  if (!RESEARCH_ROLES.has(role) && !RESEARCH_DOCX_ROLES.has(role)) {
    throw new Error(`unknown research role: ${role}`);
  }
  return `${slugify(topicPrefix)}-${role}.${ext}`;
}

/** True if a filename should be skipped from any uploads/listings. */
export function isSkipFile(filename: string): boolean {
  if (SKIP_FILES.has(filename)) return true;
  for (const p of SKIP_PREFIXES) {
    if (filename.startsWith(p)) return true;
  }
  for (const ext of SKIP_EXTENSIONS) {
    if (filename.endsWith(ext)) return true;
  }
  return false;
}

/** Classify a filename into one of: studio | research | research-docx | skip | unknown. */
export type FileClass = "studio" | "research" | "research-docx" | "skip" | "unknown";

export function classifyFile(filename: string): FileClass {
  if (isSkipFile(filename)) return "skip";
  if (STUDIO_FILENAME_REGEX.test(filename)) return "studio";
  if (RESEARCH_FILENAME_REGEX.test(filename)) return "research";
  if (RESEARCH_DOCX_FILENAME_REGEX.test(filename)) return "research-docx";
  return "unknown";
}

/** Parse a Studio filename into its components, or null if not Studio-shaped. */
export function parseStudioFilename(
  filename: string,
): { titleSlug: string; timestamp: string; product: string; ext: string } | null {
  const m = filename.match(STUDIO_FILENAME_REGEX);
  if (!m) return null;
  return { titleSlug: m[1], timestamp: m[2], product: m[3], ext: m[4] };
}

/** Parse a research filename into its components, or null if not research-shaped. */
export function parseResearchFilename(
  filename: string,
): { topicPrefix: string; role: string; ext: string } | null {
  const m = filename.match(RESEARCH_FILENAME_REGEX);
  if (!m) return null;
  return { topicPrefix: m[1], role: m[2], ext: m[3] };
}

// ── Content-type map (Supabase Storage uploads) ─────────────────────

/**
 * File extension → MIME content-type for Supabase Storage uploads.
 * Consolidated from agent/executor.ts + agent/scripts/finalize-recovered-run.ts
 * (S34 conciseness #1). TS-only — Python pipeline doesn't upload. If a Python
 * upload path is added later, mirror this into conventions.json + conventions.py.
 */
export const CONTENT_TYPES: Record<string, string> = {
  ".json": "application/json",
  ".md": "text/markdown",
  ".txt": "text/plain",
  ".html": "text/html",
  ".pdf": "application/pdf",
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".webp": "image/webp",
  ".mp3": "audio/mpeg",
  ".wav": "audio/wav",
  ".mp4": "video/mp4",
  ".webm": "video/webm",
  ".pptx": "application/vnd.openxmlformats-officedocument.presentationml.presentation",
  ".docx": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
};

/** Look up MIME content-type for a filename; falls back to application/octet-stream. */
export function getContentType(filename: string): string {
  return CONTENT_TYPES[path.extname(filename).toLowerCase()] ?? "application/octet-stream";
}

File: agent/lib/conventions.py
# -*- coding: utf-8 -*-
"""Dynamic Research conventions — Python wrapper.

Canonical data lives in agent/lib/conventions.json. This module loads it
at runtime and exposes idiomatic Python helpers. Don't duplicate values
here — change conventions.json, the wrapper picks up.

Usage:
    import sys
    sys.path.insert(0, r"<path-to-agent>/lib")
    import conventions
    print(conventions.slugify("The Quickbase Modernization Dilemma"))
    print(conventions.studio_filename("Foo Bar", "20260507-125556", "video"))
"""
from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Optional

# ── Load canonical data ───────────────────────────────────────────────

_CONVENTIONS_PATH = Path(__file__).parent / "conventions.json"

with open(_CONVENTIONS_PATH, "r", encoding="utf-8") as _f:
    _data = json.load(_f)

# ── Public constants ──────────────────────────────────────────────────

VERSION: int = _data["_version"]
LAST_UPDATED: str = _data["_last_updated"]
BUCKET: str = _data["supabase_storage"]["bucket"]

SKIP_FILES: set[str] = set(_data["skip_files"]["exact"])
SKIP_PREFIXES: list[str] = list(_data["skip_files"]["prefixes"])
SKIP_EXTENSIONS: list[str] = list(_data["skip_files"].get("extensions", []))

RESEARCH_ROLES: set[str] = set(_data["filename_patterns"]["research"]["roles"])
RESEARCH_DOCX_ROLES: set[str] = set(_data["filename_patterns"]["research_docx_companion"]["roles"])
STUDIO_PRODUCTS: dict = _data["filename_patterns"]["studio"]["products"]

_STUDIO_REGEX = re.compile(_data["filename_patterns"]["studio"]["regex"])
_RESEARCH_REGEX = re.compile(_data["filename_patterns"]["research"]["regex"])
_RESEARCH_DOCX_REGEX = re.compile(_data["filename_patterns"]["research_docx_companion"]["regex"])

_SLUG_STRIP = re.compile(_data["slugify"]["strip_pattern_python"])
_SLUG_MAX = int(_data["slugify"]["max_length"])
_SLUG_FALLBACK = _data["slugify"]["fallback"]


# ── Public API ────────────────────────────────────────────────────────


def slugify(title: Optional[str]) -> str:
    """Canyon Lake S12 slugify: strip special, spaces->hyphens, lowercase, max length."""
    if not title:
        return _SLUG_FALLBACK
    s = _SLUG_STRIP.sub("", str(title))
    s = re.sub(r"\s+", "-", s.strip())
    s = re.sub(r"-+", "-", s).lower()
    s = s[:_SLUG_MAX].rstrip("-")
    return s or _SLUG_FALLBACK


def studio_filename(title: str, timestamp: str, product: str) -> str:
    """Build a Studio product filename: {title-slug}-{TIMESTAMP}-{product}.{ext}"""
    if product not in STUDIO_PRODUCTS:
        raise ValueError(f"unknown studio product: {product}")
    ext = STUDIO_PRODUCTS[product]["ext"]
    return f"{slugify(title)}-{timestamp}-{product}.{ext}"


def research_filename(topic_prefix: str, role: str, ext: str) -> str:
    """Build a research file filename: {topic-prefix}-{role}.{ext}"""
    if role not in RESEARCH_ROLES and role not in RESEARCH_DOCX_ROLES:
        raise ValueError(f"unknown research role: {role}")
    return f"{slugify(topic_prefix)}-{role}.{ext}"


def is_skip_file(filename: str) -> bool:
    """True if a filename should be skipped from any uploads/listings."""
    if filename in SKIP_FILES:
        return True
    for p in SKIP_PREFIXES:
        if filename.startswith(p):
            return True
    for ext in SKIP_EXTENSIONS:
        if filename.endswith(ext):
            return True
    return False


def classify_file(filename: str) -> str:
    """Classify a filename into one of: studio | research | research-docx | skip | unknown."""
    if is_skip_file(filename):
        return "skip"
    if _STUDIO_REGEX.match(filename):
        return "studio"
    if _RESEARCH_REGEX.match(filename):
        return "research"
    if _RESEARCH_DOCX_REGEX.match(filename):
        return "research-docx"
    return "unknown"


def parse_studio_filename(filename: str) -> Optional[dict]:
    """Parse a Studio filename, or None if not Studio-shaped."""
    m = _STUDIO_REGEX.match(filename)
    if not m:
        return None
    return {"title_slug": m.group(1), "timestamp": m.group(2), "product": m.group(3), "ext": m.group(4)}


def parse_research_filename(filename: str) -> Optional[dict]:
    """Parse a research filename, or None if not research-shaped."""
    m = _RESEARCH_REGEX.match(filename)
    if not m:
        return None
    return {"topic_prefix": m.group(1), "role": m.group(2), "ext": m.group(3)}


# ── Self-test when run directly ───────────────────────────────────────

if __name__ == "__main__":
    print(f"Conventions v{VERSION} ({LAST_UPDATED})")
    print(f"BUCKET: {BUCKET}")
    print(f"slugify('The Quickbase Modernization Dilemma: Domain Specificity vs. Infrastructure Security'):")
    print(f"  -> {slugify('The Quickbase Modernization Dilemma: Domain Specificity vs. Infrastructure Security')}")
    print(f"studio_filename('Foo Bar Baz', '20260507-125556', 'video'):")
    print(f"  -> {studio_filename('Foo Bar Baz', '20260507-125556', 'video')}")
    print(f"research_filename('cam-ai-quickbase-platform', 'brief', 'md'):")
    print(f"  -> {research_filename('cam-ai-quickbase-platform', 'brief', 'md')}")
    print(f"is_skip_file('job-manifest.json'): {is_skip_file('job-manifest.json')}")
    print(f"is_skip_file('helper.py'): {is_skip_file('helper.py')}")
    print(f"is_skip_file('nlm-discovered.json'): {is_skip_file('nlm-discovered.json')}")
    print(f"classify_file('the-foo-20260507-125556-video.mp4'): {classify_file('the-foo-20260507-125556-video.mp4')}")
    print(f"classify_file('cam-ai-brief.md'): {classify_file('cam-ai-brief.md')}")
    print(f"classify_file('cam-ai-context.md'): {classify_file('cam-ai-context.md')}")
    print(f"classify_file('random.txt'): {classify_file('random.txt')}")
